collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java

```diff
@@ -265,269 +169,102 @@ private void deleteFileOrDir(String fileName) {
-                            }
-
-                            if (values.containsKey(FormsColumns.FORM_FILE_PATH)) {
-                                String formFile = storagePathProvider.getAbsoluteFormFilePath(values
-                                        .getAsString(FormsColumns.FORM_FILE_PATH));
-                                String oldFile = storagePathProvider.getAbsoluteFormFilePath(update.getString(update
-                                        .getColumnIndex(FormsColumns.FORM_FILE_PATH)));
-
-                                if (formFile == null || !formFile.equalsIgnoreCase(oldFile)) {
-                                    deleteFileOrDir(oldFile);
-                                }
-
-                                // we're updating our file, so update the md5
-                                // and get rid of the cache (doesn't harm anything)
-                                deleteFileOrDir(storagePathProvider.getAbsoluteCacheFilePath(update
-                                        .getString(update
-                                                .getColumnIndex(FormsColumns.JRCACHE_FILE_PATH))));
-                                String newMd5 = FileUtils
-                                        .getMd5Hash(new File(formFile));
-                                values.put(FormsColumns.MD5_HASH, newMd5);
-                                values.put(FormsColumns.JRCACHE_FILE_PATH,
-                                        storagePathProvider.getRelativeCachePath(newMd5 + ".formdef"));
-                            }
-
-                            count = db.update(
-                                    FORMS_TABLE_NAME,
-                                    values,
-                                    FormsColumns._ID
-                                            + "=?"
-                                            + (!TextUtils.isEmpty(where) ? " AND ("
-                                            + where + ')' : ""), prepareWhereArgs(whereArgs, formId));
-                        } else {
-                            Timber.e("Attempting to update row that does not exist");
-                        }
-                    } finally {
-                        if (update != null) {
-                            update.close();
-                        }
+                        formsRepository.save(getFormFromValues(existingValues, storagePathProvider));
                     }
-                    break;
 
-                default:
-                    throw new IllegalArgumentException("Unknown URI " + uri);
-            }
+                    count = cursor.getCount();
+                }
+                break;
 
-            getContext().getContentResolver().notifyChange(uri, null);
-            getContext().getContentResolver().notifyChange(FormsColumns.CONTENT_NEWEST_FORMS_BY_FORMID_URI, null);
-        }
+            case FORM_ID:
+                Form form = formsRepository.get(ContentUriHelper.getIdFromUri(uri));
```

Comment:
You may need a null check below?

Reply:
Hmmm good question. I'll look at what the previous behaviour was (if a non-existent ID was passed in) and write a test to make sure this implementation is consistent.